.idea/readme.md

```diff
@@ -30,6 +30,17 @@ _**PyCharm’s remote debugger requires PyCharm Professional._
 
 <br />
 
+<h1>Setting up Python Code Completion</h1>
+
+In Pycharm or Rider, after installing [`lean-cli`](https://github.com/QuantConnect/lean-cli) or [`quantconnect-stubs`](https://github.com/QuantConnect/quantconnect-stubs-generator),
+
+1. Open settings
+2. Click on "Python Console" under the "Console" subsection
+3. For the interpreter of your choice, add [the required imports](https://github.com/QuantConnect/Lean/blob/master/Algorithm.Python/readme.md#python-autocomplete-imports) to the bottom of "Starting script"
+
+After clicking "Apply" or "Ok", all .py files opened using these project settings will implicitly import the stubs module. 
+It is recommended that a virtual environment is used for your Lean project so that these stubs will only ever load when developing in Lean.
```

Comment:
I dug a little deeper into what the "Starting script" does, commenting on the commit instead of the old thread because this is the only way to add a suggestion.

> I'll change the recommended imports to paste to a link to the part of the ReadMe which lists these out.

Sounds good, but the "Starting script" is the wrong place to put those imports. If you want autocompletion in your .py files you'll need to place those imports in your .py file itself, since the "Starting script" only contains the code that needs to run when a Python console is opened (i.e. it doesn't affect the autocompletion in .py files).

In case the above is unclear, this is how I would write it:

```suggestion
<h1>Setting up Python Code Completion</h1>

In Pycharm or Rider, after installing [`lean-cli`](https://github.com/QuantConnect/lean-cli) or [`quantconnect-stubs`](https://github.com/QuantConnect/quantconnect-stubs-generator):

1. Open the settings (File > Settings).
2. Click on "Python Interpreter" under the "Project: <project name>" subsection.
3. Set the Python Interpreter to the one that you used to install the CLI or the stubs.

After clicking "Apply" or "Ok", all .py files opened using these project settings will be aware of the `QuantConnect.*` and `System.*` namespaces, and the IDE will provide autocompletion on them. Copy the imports listed [here](https://github.com/QuantConnect/Lean/blob/master/Algorithm.Python/readme.md#python-autocomplete-imports) into your .py files to import everything that is automatically provided to Python algorithms on QuantConnect.com.

It is recommended that a virtual environment is used for your Lean project so that these stubs will only ever load when developing in Lean.
```

> For the installation source, lean-cli being installed would imply that quantconnect-stubs would also be given that `install_requires` has the stubs as a dependency -- I figured I'd mention either should satisfy this.

:+1:

> Lastly, for the error, I was getting it too found a non-elegant way to fix it. In my case, looking into my clr module, it had been replaced by stub files (i.e., clr only contained `__init__.pyi`). Having it there was actually preventing pythonnet from being installed too -- in my case, `rm -r (...your dir)/site-packages/clr` and then installing pythonnet and the stubs allowed for importing as expected.
> It seems the stubs were put there by the generator -- perhaps pythonnet needs to be made a hard dependency for the generator to avoid this.

The error seems to have to do with the code being ran in the wrong place. You're not supposed to run "import QuantConnect" in a Python console, executing such an import will only work when it is part of an algorithm ran using Lean. When running a Python algorithm with the Lean launcher this error should not happen.

Reply:
>The error seems to have to do with the code being ran in the wrong place. You're not supposed to run "import QuantConnect" in a Python console, executing such an import will only work when it is part of an algorithm ran using Lean. When running a Python algorithm with the Lean launcher this error should not happen.

Good point. Admittedly, I wasn't using the console when developing locally for this reason but wanted the stubs available immediately without needing to manually copy + paste into every .py. Of course, this ended up sort of ruining the interpreter when used as a console. I think its worth seeing if this can be achieved by some other means before reopening the pr, though.

Reply:
I'm not exactly sure how the stubs behaved when they were included in the repo, but I believe you still needed to have the actual imports in the .py files right? If you leave out `from QuantConnect.Algorithm import QCAlgorithm` Lean will throw an error when you run the algorithm.

I do agree that copying and pasting 20+ imports into every .py file isn't very nice, but since you're using a JetBrains IDE you can just start typing the type that you need (like "QCAlgorithm") and when you're near the end just hit Ctrl+Space twice and the IDE will show you a nice pop-up with all imports containing the name that you're typing:

![image](https://user-images.githubusercontent.com/14951909/108795984-f2d4d000-7587-11eb-93bb-2b035694cd18.png)

Reply:
IIRC, when the stubs were included, by specifying the folder as a namespace allowed for parsing as if all of the imports were included.

Reply:
Do you mean you are copying all the stubs from the `quantconnect-stubs` package into your project root?